DietPi-DDNS | Add provider YDNS

TDuffinNTU opened the pull request:
Closes #5128 

As YDNS expects a single (one and only one) domain in its API request I've added a special case for it in Menu_Domains but this might not be correct (do we submit domains as a  _list_ or is it split by the commas and sent in multiple requests?

There are two options for YDNS updating - One uses an auth header and the other uses an "update link" which is a shortened variant of the update URI but with a random token. I opted for the former since I feel that's more secure and easier to fit into the UX of the application. 

Remaining in draft until I can spin up a dietpi VM to test this, but welcome to feedback ahead of that.

TDuffinNTU committed:
add support for YDNS

TDuffinNTU committed:
remove TODO

MichaIng commented on dietpi/dietpi-ddns:
```diff
@@ -40,6 +40,8 @@ Available providers:
   OVH               Read more: https://docs.ovh.com/gb/en/domains/hosting_dynhost/
                     Use the "-d", "-u" and "-p" options to set domains, username and password.
+  YDNS				Read More https://ydns.io
+					Use the "-d", "-u" and "-p" options to set domains, username and password.
```
We use regular blank spaces here to assure correct alignment. And if only one domain is allowed, let's make that clear here as well:
```suggestion
                    Use the "-d", "-u" and "-p" options to set domains, username and password.
  YDNS              Read More https://ydns.io
                    Use the "-d", "-u" and "-p" options to set the domain, username and password.
```

MichaIng commented on dietpi/dietpi-ddns:
```diff
@@ -290,6 +306,7 @@ Menu_Provider()
 		'Dynu' ': Read more: https://www.dynu.com/DynamicDNS'
 		'FreeDNS' ': Read more: https://freedns.afraid.org/'
 		'OVH' ': Read More: https://docs.ovh.com/gb/en/domains/hosting_dynhost/'
+		'YDNS' ': Read More https://ydns.io'
```
```suggestion
		'OVH' ': Read more: https://docs.ovh.com/gb/en/domains/hosting_dynhost/'
		'YDNS' ': Read more: https://ydns.io'
```

MichaIng commented on dietpi/dietpi-ddns:
```diff
@@ -122,6 +124,14 @@ Read()
 		[[ $PASSWORD ]] || PASSWORD=${command#*:} PASSWORD=${PASSWORD%%\'*}
 
+	# YDNS
+	elif [[ $command = *'www.ydns.io'*]]
```
Syntax
```suggestion
	elif [[ $command == *'www.ydns.io'* ]]
```

MichaIng commented on dietpi/dietpi-ddns:
```diff
@@ -302,22 +319,29 @@ Menu_Provider()
 
 	G_WHIP_DEFAULT_ITEM=$DOMAINS
-	G_WHIP_INPUTBOX 'Please enter a comma-separated list of domains that shall point to this system:' || return 1
+	
```
Trailing space, actually the empty line can probably be removed, since the `G_WHIP_DEFAULT_ITEM` belong to those input boxes anyway:
```suggestion
```

TDuffinNTU committed:
Apply suggestions from code review

Syntax/typo/spelling corrections.

Co-authored-by: MichaIng <micha@dietpi.com>

TDuffinNTU committed:
typo causing ddns script to miss domains menu

TDuffinNTU committed:
Swapped conditional around to make code a bit more clear (think this was also actual reason YDNS was skipping its domains menu)

MichaIng commented on dietpi/dietpi-ddns:
```diff
@@ -122,6 +124,14 @@ Read()
+	elif [[ $command == *'www.ydns.io'* ]]
+	then
+		[[ $PROVIDER ]] || PROVIDER='YDNS'
+		[[ $DOMAINS ]] || DOMAINS=${command#*hostname=} DOMAINS=${DOMAINS%\'*}
```
```suggestion
		[[ $DOMAINS ]] || DOMAINS=${command#*host=} DOMAINS=${DOMAINS%\'*}
```

MichaIng commented on dietpi/dietpi-ddns:
```diff
@@ -307,17 +324,23 @@ Menu_Provider()
-	# Skip with FreeDNS and custom provider
-	[[ $PROVIDER == 'DuckDNS' || $PROVIDER == 'No-IP' || $PROVIDER == 'Dynu' || $PROVIDER == 'OVH' ]] || return 0
+	# Skip with these providers
+	[[ $PROVIDER == 'FreeDNS' ]] && return 0
```
It is not beautiful, but let's keep the previous method, otherwise the menu is not skipped with custom providers:
```suggestion
	# Skip with FreeDNS and custom provider
	[[ $PROVIDER == 'DuckDNS' || $PROVIDER == 'No-IP' || $PROVIDER == 'Dynu' || $PROVIDER == 'OVH' || $PROVIDER == 'YDNS' ]] || return 0
```
When one finds time, we should create some clear arrays at the start of the script which contain all information for each provider, whether it supports domain, username, password and/or token input, the API URL and such. So adding a new provider could be done with just adding array entries in one place in the code and the menus are correctly generated based on the arrays' content. Then we can also get rid of these growing conditions in the code.

TDuffinNTU replied:
Fair enough! Agreed that it would be good to refactor this especially since misspelling an entry across the file can easily cause a tricky bug.

TDuffinNTU replied:
Applied the change to my latest commit, but added YDNS to the list too :)

TDuffinNTU committed:
reversed change, but added ydns to list of providers that should NOT be skipped

MichaIng commented on dietpi/dietpi-ddns:
```diff
@@ -307,17 +324,23 @@ Menu_Provider()
-	# Skip with FreeDNS and custom provider
-	[[ $PROVIDER == 'DuckDNS' || $PROVIDER == 'No-IP' || $PROVIDER == 'Dynu' || $PROVIDER == 'OVH' ]] || return 0
+	# Skip with FreeDNS/Custom providers
+	[[ $PROVIDER == 'DuckDNS' || $PROVIDER == 'No-IP' || $PROVIDER == 'Dynu' || $PROVIDER == 'OVH' || $Provider == 'YDNS' ]] || return 0
```
Capital letters 🙂:
```suggestion
	[[ $PROVIDER == 'DuckDNS' || $PROVIDER == 'No-IP' || $PROVIDER == 'Dynu' || $PROVIDER == 'OVH' || $PROVIDER == 'YDNS' ]] || return 0
```

TDuffinNTU committed:
Update dietpi/dietpi-ddns

Co-authored-by: MichaIng <micha@dietpi.com>

TDuffinNTU committed:
Update dietpi/dietpi-ddns

Co-authored-by: MichaIng <micha@dietpi.com>

MichaIng committed:
v8.23

- CHANGELOG | DietPi-DDNS: YDNS has been added to the list of natively supported DDNS providers. Many thanks to @edmundlaugasson for requesting and @TDuffinNTU for implementing it: https://github.com/MichaIng/DietPi/issues/5128, https://github.com/MichaIng/DietPi/pull/6674

TDuffinNTU committed:
Merge branch 'dev' into dev
